src/main/java/de/tum/cit/aet/artemis/programming/service/AutomaticProgrammingExerciseCleanupService.java

```diff
@@ -82,49 +89,66 @@ public void cleanup() {
-        log.info("Search for exercises with course or exam date from {} until {}", endDate1, endDate2);
-        programmingExercises = programmingExerciseRepository.findAllByRecentCourseEndDate(endDate1, endDate2);
-        programmingExercises.addAll(programmingExerciseRepository.findAllByRecentExamEndDate(endDate1, endDate2));
+        log.info("Search for exercises with course or exam date from {} until {}", earliestDate, latestDate);
+        var programmingExercises = programmingExerciseRepository.findAllByRecentCourseEndDate(earliestDate, latestDate);
+        programmingExercises.addAll(programmingExerciseRepository.findAllByRecentExamEndDate(earliestDate, latestDate));
         log.info("Found {} programming exercise to clean local template, test and solution: {}", programmingExercises.size(),
                 programmingExercises.stream().map(ProgrammingExercise::getProjectKey).collect(Collectors.joining(", ")));
-        for (var programmingExercise : programmingExercises) {
-            gitService.deleteLocalRepository(programmingExercise.getVcsTemplateRepositoryUri());
-            gitService.deleteLocalRepository(programmingExercise.getVcsSolutionRepositoryUri());
-            gitService.deleteLocalRepository(programmingExercise.getVcsTestRepositoryUri());
-            gitService.deleteLocalProgrammingExerciseReposFolder(programmingExercise);
+        if (!programmingExercises.isEmpty()) {
+            for (var programmingExercise : programmingExercises) {
+                gitService.deleteLocalRepository(programmingExercise.getVcsTemplateRepositoryUri());
+                gitService.deleteLocalRepository(programmingExercise.getVcsSolutionRepositoryUri());
+                gitService.deleteLocalRepository(programmingExercise.getVcsTestRepositoryUri());
+                gitService.deleteLocalProgrammingExerciseReposFolder(programmingExercise);
+            }
+            log.info("Finished cleaning local template, test and solution repositories");
+        }
+    }
+
+    private void cleanStudentParticipationsRepositories(ZonedDateTime earliestDate, ZonedDateTime latestDate) {
+        log.info("Search for exercises with due date from {} until {}", earliestDate, latestDate);
+        // Get all relevant participation ids
+        Pageable pageable = Pageable.ofSize(STUDENT_PARTICIPATION_CLEANUP_BATCH_SIZE);
+        Page<String> uriBatch = programmingExerciseStudentParticipationRepository.findRepositoryUrisForGitCleanupByRecentDueDateOrRecentExamEndDate(earliestDate, latestDate,
+                pageable);
+        log.info("Found {} student participations to clean local student repositories in {} batches.", uriBatch.getTotalElements(), uriBatch.getTotalPages());
+        if (uriBatch.getTotalElements() > 0) {
+            var batchStream = Stream.iterate(uriBatch, Page::hasNext, page -> programmingExerciseStudentParticipationRepository
```

Comment:
The `Page::hasNext` is checked before the current page is included in the Stream. The last Page has no next Page so it is not included.

Reply:
Thank you! The implementation of the iterate method and the hasNext of pages made it kinda confusing for me.